aiagents4pharma/talk2scholars/tests/test_s2_search.py

```diff
@@ -0,0 +1,259 @@
+    dummy_requests_get_exception.called_timeout = timeout
+    # Simulate a network/connection exception.
+    raise requests.exceptions.RequestException("Connection error")
+
+
+# --- Pytest Fixture to Patch Hydra ---
+@pytest.fixture(autouse=True)
+def patch_hydra(monkeypatch):
+    """hydra patch to mock initialize and compose functions."""
+    # Patch hydra.initialize to return our dummy context manager.
+    monkeypatch.setattr(
+        hydra, "initialize", lambda version_base, config_path: DummyHydraContext()
+    )
+    # Patch hydra.compose to return our dummy config.
+    monkeypatch.setattr(hydra, "compose", lambda config_name, overrides: dummy_config)
+
+
+# --- Test Cases ---
+
+
+def test_search_tool_success(monkeypatch):
+    """
+    Test that search_tool returns a valid Command object when the API response is successful.
+    Also checks that papers without required fields are filtered out.
+    """
+    monkeypatch.setattr(requests, "get", dummy_requests_get_success)
+
+    tool_call_id = "test_tool_call_id"
+    # Invoke using .run() with a dictionary input.
+    result = search_tool.run(
+        {
+            "query": "machine learning",
+            "tool_call_id": tool_call_id,
+            "limit": 3,
+            "year": "2020",
+        }
+    )
+
+    # Check that a Command is returned with the expected update structure.
+    assert isinstance(result, Command)
+    update = result.update
+    assert "papers" in update
+
+    papers = update["papers"]
+    # Papers with valid 'title' and 'authors' should be included.
+    assert "1" in papers
+    assert "2" in papers
+    # Paper "3" is missing authors and should be filtered out.
+    assert "3" not in papers
+
+    # Check that a ToolMessage is included in the messages.
+    messages = update.get("messages", [])
+    assert len(messages) == 1
+    msg = messages[0]
+    assert isinstance(msg, ToolMessage)
+    assert "Number of papers found:" in msg.content
+
+    # Verify that the correct parameters were sent to requests.get.
+    called_params = dummy_requests_get_success.called_params
+    assert called_params["query"] == "machine learning"
+    # The "year" parameter should have been added.
+    assert called_params["year"] == "2020"
+    # The limit is set to min(limit, 100) so it should be 3.
+    assert called_params["limit"] == 3
+    # The fields should be a comma-separated string from the dummy config.
+    assert called_params["fields"] == "paperId,title,authors"
+
+
+def test_search_tool_unexpected_format(monkeypatch):
+    """
+    Test that search_tool raises a RuntimeError when the API response
+    does not include the expected 'data' key.
+    """
+    monkeypatch.setattr(requests, "get", dummy_requests_get_no_data)
+    tool_call_id = "test_tool_call_id"
+    with pytest.raises(
+        RuntimeError,
+        match="Unexpected response from Semantic Scholar API. Please retry the same query.",
```

Comment:
If there is an error in the formatting (or no paper returned) should the agent retry the same query or would it better to let the user know and then modify their query accordingly?